third_party/move/move-compiler-v2/src/file_format_generator/peephole_optimizer/optimizers.rs

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module contains setup for basic block peephole optimizers.
+
+use move_binary_format::file_format::Bytecode;
+use std::mem;
+
+/// A basic block optimizer that optimizes a basic block of bytecode.
+pub trait BasicBlockOptimizer {
+    /// Given a basic `block`, return its optimized version.
+    fn optimize(&self, block: &[Bytecode]) -> Vec<Bytecode>;
+}
+
+/// An optimizer for a fixed window of bytecode.
+/// The fixed window can be assumed to be within a basic block.
+pub trait FixedWindowOptimizer {
+    /// The fixed window size for this optimizer.
+    fn fixed_window_size(&self) -> usize;
+
+    /// Given a fixed `window` of bytecode of size `self.fixed_window_size()`,
+    /// optionally return its optimized version.
+    /// If `None` is returned, the `window` is not optimized.
+    fn optimize_fixed_window(&self, window: &[Bytecode]) -> Option<Vec<Bytecode>>;
+}
+
+/// A processor to perform fixed window optimizations of a particular style on a basic block.
+pub struct FixedWindowProcessor<T: FixedWindowOptimizer>(T);
+
+impl<T: FixedWindowOptimizer> BasicBlockOptimizer for FixedWindowProcessor<T> {
+    fn optimize(&self, block: &[Bytecode]) -> Vec<Bytecode> {
+        let mut old_block = block.to_vec();
+        // Run single passes until code stops changing.
+        while let Some(mut new_block) = self.optimize_single_pass(&old_block) {
+            old_block = mem::take(&mut new_block);
```

Comment:
Note that `Vec` doesn't implement `Copy`, so you should be able to just say
```
    old_block = new_block;
```
here.

Reply:
Nice catch, done.